src/main/java/com/vlkan/pubsub/PubsubAccessTokenCache.java

```diff
@@ -0,0 +1,136 @@
+        builder.executorService.scheduleAtFixedRate(
+                this::refreshAccessToken,
+                accessTokenRefreshPeriodMillis,
+                accessTokenRefreshPeriodMillis,
+                TimeUnit.MILLISECONDS);
+    }
+
+    String getAccessToken() {
+        return accessToken;
+    }
+
+    private static GoogleCredentials createGoogleCredentials(@Nullable String credentials) {
+        GoogleCredentials googleCredentials;
+        if (credentials == null) {
+            try {
+                googleCredentials = GoogleCredentials.getApplicationDefault();
+            } catch (IOException error) {
+                throw new RuntimeException("cannot retrieve default credentials", error);
+            }
+        } else {
+            Objects.requireNonNull(credentials, "credentials");
+            googleCredentials = createServiceAccountCredentials(credentials);
+        }
+        return googleCredentials.createScopedRequired()
+                ? googleCredentials.createScoped(SCOPES)
+                : googleCredentials;
+    }
+
+    private static ServiceAccountCredentials createServiceAccountCredentials(String credentials) {
+        byte[] credentialsBytes = Base64.getDecoder().decode(credentials);
+        try (InputStream secretBytesStream = new ByteArrayInputStream(credentialsBytes)) {
+            return ServiceAccountCredentials.fromStream(secretBytesStream);
+        } catch (IOException error) {
+            throw new RuntimeException(error);
+        }
+    }
+
+    private void refreshAccessToken() {
+        try {
+            credentials.refreshIfExpired();
+            this.accessToken = credentials.getAccessToken().getTokenValue();
+        } catch (IOException error) {
+            LOGGER.error("Failed to fetch access token", error);
+        }
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Builder {
+
+        private ScheduledExecutorService executorService;
+
+        private Duration accessTokenRefreshPeriod = DEFAULT_ACCESS_TOKEN_REFRESH_PERIOD;
+
+        @Nullable
+        private String credentials;
+
+        private Builder() {}
+
+        public Builder setExecutorService(ScheduledExecutorService executorService) {
+            this.executorService = Objects.requireNonNull(executorService, "executorService");
+            return this;
+        }
+
+        public Builder setAccessTokenRefreshPeriod(Duration accessTokenRefreshPeriod) {
+            this.accessTokenRefreshPeriod = Objects.requireNonNull(accessTokenRefreshPeriod, "accessTokenRefreshPeriod");
+            return this;
+        }
+
+        public Builder setCredentials(@Nullable String credentials) {
+            this.credentials = credentials;
+            return this;
+        }
+
+        public PubsubAccessTokenCache build() {
+            Objects.requireNonNull(executorService, "executorService");
```

Comment:
A random thought:

Have you considered using type-safe builders? To enforce the users to provide the required settings. Something like:
```java
class Builder {
// ...
}

class ExecutorServiceBuilder {
    Builder executorService(...) {}
}

class PubsubAccessTokenCache {
    public ExecutorServiceBuilder builder() { /* ... */ }
}
```

Reply:
Yes, I've thought of implementing [staged/telescopic builders](http://immutables.github.io/immutable.html#staged-builder), though moved away from that idea due to two concerns:

1. They are difficult to evolve without breaking backward compatibility.
2. Given builders are a work around for Java's lack of named parameters, I don't want to stretch this hack more than the bare minimum.

If you think my concerns are superficial (which very well can be), I might revisit the idea.